### ISSUE TITLE
addMethod yields ts error 1.0.0

### ISSUE THREAD

**Issue by walidvb** [jquense/yup#1899]
**Describe the bug**
When using addMethods as per the readme, a ts error is thrown

Message: 

```
No overload matches this call.
  Overload 1 of 2, '(schemaType: (...arg: any[]) => AnySchema<any, any, any, Flags>, name: string, fn: (this: AnySchema<any, any, any, Flags>, ...args: any[]) => AnySchema<any, any, any, Flags>): void', gave the following error.
    Argument of type 'typeof create$6' is not assignable to parameter of type '(...arg: any[]) => AnySchema<any, any, any, Flags>'.
      Call signature return types 'StringSchema<string | undefined, AnyObject, undefined, "">' and 'AnySchema<any, any, any, Flags>' are incompatible.
        The types of '__outputType' are incompatible between these types.
          Type 'string | undefined' is not assignable to type 'never'.
            Type 'undefined' is not assignable to type 'never'.
  Overload 2 of 2, '(schemaType: new (...args: any) => AnySchema<any, any, any, Flags>, name: string, fn: (this: AnySchema<any, any, any, Flags>, ...args: any[]) => AnySchema<any, any, any, Flags>): void', gave the following error.
    Argument of type 'typeof create$6' is not assignable to parameter of type 'new (...args: any) => AnySchema<any, any, any, Flags>'.
      Type 'typeof create$6' provides no match for the signature 'new (...args: any): AnySchema<any, any, any, Flags>'.ts(2769)

https://z0kq0k.csb.app/
```

**To Reproduce**

https://codesandbox.io/s/ancient-glade-z0kq0k?file=/src/index.ts

> NOTE: if you do not provide a runnable reproduction the chances of getting feedback are significantly lower

**Expected behavior**
Typescript accepts

**Platform (please complete the following information):**
 - Browser any
 - Version any

**Comment by walidvb** [comment]
@jquense congrats on the v1 release?
Have you had a chance to look at this already? are you not experiencing this isue?

**Comment by afewyards** [comment]
I'm having the same issue with this. preventing me to upgrade to 1.0.0

**Comment by NagaiKoki** [comment]
I think the reason of this type error is the `addMethod`'s  type parameter `"AnySchema"` doesn't exist `declare class AnySchema`, but just only `interface AnySchema`. (There are two `StringSchema` types in `index.d.ts`, `declare class StringSchema` and `interface StringSchema`.  So `AnySchema` must have declare class type.)

```ts
declare function addMethod<T extends AnySchema>(schemaType: (...arg: any[]) => T, name: string, fn: (this: T, ...args: any[]) => T): void;
type AnySchema<TType = any, C = any, D = any, F extends Flags = Flags> = Schema<TType, C, D, F>;

declare class StringSchema<TType extends Maybe<string> = string | undefined, TContext = AnyObject, TDefault = undefined, TFlags extends Flags = ''> extends Schema<TType, TContext, TDefault, TFlags> {
    constructor();
    length(length: number | Reference<number>, message?: Message<{
        length: number;
    }>): this;
   // ...
}
interface StringSchema<TType extends Maybe<string> = string | undefined, TContext = AnyObject, TDefault = undefined, TFlags extends Flags = ''> extends Schema<TType, TContext, TDefault, TFlags> {
    default<D extends Maybe<TType>>(def: Thunk<D>): StringSchema<TType, TContext, D, ToggleDefault<TFlags, D>>;
     // ...
}
```

**Comment by NagaiKoki** [comment]
I look into this issue a bit deeply, and turned out this is occured by the `__outputType` type of `abstract class Schema` class.
First, `StringSchema` type which is passed by `addMethod`'s default type parameter and `Schema` type which is extended by this are below.

```ts
export default class StringSchema<
  TType extends Maybe<string> = string | undefined,
  TContext = AnyObject,
  TDefault = undefined,
  TFlags extends Flags = '',
> extends Schema<TType, TContext, TDefault, TFlags>

export default abstract class Schema<
  TType = any,
  TContext = any,
  TDefault = any,
  TFlags extends Flags = '',
> implements ISchema<TType, TContext, TFlags, TDefault>
{
  readonly type: string;
  declare readonly __outputType: ResolveFlags<TType, TFlags, TDefault>;
```

In this case, the `__outputType`'s type is output as `any`.

Second, `AnySchema` type which is `addMethod`'s extend type parameter is below.

```ts
export type AnySchema<
  TType = any,
  C = any,
  D = any,
  F extends Flags = Flags,
> = Schema<TType, C, D, F>;
```

In this case, `__outputType`'s type which is `Schema` class is output as `never`.

In summary, the type of `__outputType` generated by `StringSchema` (`any`) is not subtype of `AnySchema`'s `_outputType` type (`never`), so `addMethod<StringSchema extends AnySchema>` is occured type error.
I'm not sure how to fix of it for now, and the workaround I've just noted above is just to overload exist `AnySchema`'s type and enabled `StringSchema` to subtype of `AnySchema`.

**Comment by jquense** [comment]
fixed in 1.0.1
